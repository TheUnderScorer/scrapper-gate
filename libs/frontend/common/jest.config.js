module.exports = {
  displayName: 'frontend-common',
  preset: '../../../jest.preset.js',
  transform: {
    '^.+\\.[tj]sx?$': 'ts-jest',
  },
  moduleFileExtensions: ['ts', 'tsx', 'js', 'jsx'],
  coverageDirectory: '../../../coverage/libs/frontend/common',
};
