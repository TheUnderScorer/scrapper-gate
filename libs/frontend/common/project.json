{
  "root": "libs/frontend/common",
  "sourceRoot": "libs/frontend/common/src",
  "projectType": "library",
  "targets": {
    "lint": {
      "executor": "@nrwl/linter:eslint",
      "options": {
        "lintFilePatterns": ["libs/frontend/common/**/*.{ts,tsx,js,jsx}"]
      },
      "outputs": ["{options.outputFile}"]
    },
    "test": {
      "executor": "@nrwl/jest:jest",
      "outputs": ["coverage/libs/frontend/common"],
      "options": {
        "jestConfig": "libs/frontend/common/jest.config.js",
        "passWithNoTests": true
      }
    }
  },
  "tags": []
}
