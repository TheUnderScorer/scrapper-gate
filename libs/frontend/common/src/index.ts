export * from './types/apollo';
export * from './types/selection';
export * from './stores/useAppType';
export * from './types/appType';
export * from './stores/useContainerStore';
export * from './types/components';
export * from './providers/QueryParamProvider';
export * from './os';
export * from './stopPropagation';
export * from './preventDefault';
export * from './isElementEditable';
export * from './highlight';
export * from './hooks/useLongPressValue';
export * from './hooks/useIsUsingElementPicker';
export * from './elementNavigator';
export * from './setRefValue';
export * from './stores/useCurrentUrl';
export * from './hooks/useStack';
export * from './errors';
export * from './hooks/useLocationHref';
export * from './hooks/useReturnUrlProvider';
export * from './openState/useMemoryOpenState';
export * from './openState/useQueryParamOpenState';
export * from './openState/types';
