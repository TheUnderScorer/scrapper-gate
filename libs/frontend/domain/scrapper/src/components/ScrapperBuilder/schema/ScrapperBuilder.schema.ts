import { ScrapperBuilderFormState } from '../ScrapperBuilder.types';
import {
  ScrapperRunSettingsInputSchema,
  VariableInputSchema,
} from '@scrapper-gate/shared/validation';
import joi from 'joi';
import { ScrapperStepNodeSchema } from './ScrapperStepNode.schema';

export const ScrapperBuilderSchema = joi.object<ScrapperBuilderFormState>({
  items: joi.array().items(ScrapperStepNodeSchema),
  name: joi.string().required(),
  variables: joi.array().items(VariableInputSchema).allow(null),
  runSettings: ScrapperRunSettingsInputSchema,
});
