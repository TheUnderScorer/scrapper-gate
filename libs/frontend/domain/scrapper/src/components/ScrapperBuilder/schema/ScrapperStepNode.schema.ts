import { FlowBuilderNodeTypes } from '@scrapper-gate/frontend/flow-builder';
import { ScrapperStepInputSchema } from '@scrapper-gate/shared/validation';
import joi from 'joi';
import { isEdge } from 'react-flow-renderer';
import { ScrapperBuilderNode } from '../ScrapperBuilder.types';

const omitTypes = [FlowBuilderNodeTypes.Start, FlowBuilderNodeTypes.End];

// eslint-disable-next-line @typescript-eslint/no-explicit-any
const whenSchema = joi.object().custom((item: ScrapperBuilderNode, helpers) => {
  if (isEdge(item) || omitTypes.includes(item.type as FlowBuilderNodeTypes)) {
    return helpers.error('any');
  }

  return item;
});
export const ScrapperStepNodeSchema = joi
  .object<Pick<ScrapperBuilderNode, 'data' | 'type'>>({
    data: joi.object(),
    type: joi.string(),
  })
  .when(whenSchema, {
    then: {
      data: ScrapperStepInputSchema,
    },
  });
