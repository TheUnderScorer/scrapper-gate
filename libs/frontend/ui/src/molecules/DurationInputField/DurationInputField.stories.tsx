import { Duration } from '@scrapper-gate/shared/common';
import { useState } from 'react';
import 'reflect-metadata';
import { DurationInputField } from './DurationInputField';

export default {
  title: 'UI/Duration Input Field',
};

export const Component = () => {
  const [value, setValue] = useState(Duration.fromMinutes(60).toInput());

  return (
    <>
      <DurationInputField label="Duration" value={value} onChange={setValue} />
      <pre>{JSON.stringify(value, null, ' ')}</pre>
    </>
  );
};
