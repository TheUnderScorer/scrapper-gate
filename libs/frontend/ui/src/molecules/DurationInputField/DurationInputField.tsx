import { InputAdornment, MenuItem, Select, TextField } from '@mui/material';
import {
  Duration,
  saveParseFloat,
  toDisplayText,
} from '@scrapper-gate/shared/common';
import { DurationUnit } from '@scrapper-gate/shared/schema';
import React, { useMemo } from 'react';
import { DurationInputFieldProps } from './DurationInputField.types';

const units = Object.values(DurationUnit).map((unit) => ({
  value: unit,
  label: toDisplayText(unit),
}));

export const DurationInputField = ({
  value,
  onChange,
  disabledUnits,
  ...rest
}: DurationInputFieldProps) => {
  const supportedUnits = useMemo(() => {
    if (!disabledUnits?.length) {
      return units;
    }

    return units.filter((unit) => !disabledUnits.includes(unit.value));
  }, [disabledUnits]);

  const inputValue = useMemo(() => {
    if (value && 'value' in value) {
      return value;
    }

    const newValue = value
      ? Duration.fromDuration(value).toInput()
      : Duration.fromMs(0).toInput();

    onChange?.(newValue);

    return newValue;
  }, [onChange, value]);

  return (
    <TextField
      {...rest}
      value={inputValue?.value}
      onChange={(event) =>
        onChange?.({
          value: saveParseFloat(event.target.value),
          unit: inputValue?.unit ?? DurationUnit.Milliseconds,
        })
      }
      InputProps={{
        endAdornment: (
          <InputAdornment position="end">
            <Select
              variant="standard"
              value={inputValue?.unit ?? DurationUnit.Milliseconds}
              onChange={(event) => {
                onChange?.({
                  value: inputValue?.value ?? 0,
                  unit: event.target.value as DurationUnit,
                });
              }}
            >
              {supportedUnits.map((unit) => (
                <MenuItem key={unit.value} value={unit.value}>
                  {unit.label}
                </MenuItem>
              ))}
            </Select>
          </InputAdornment>
        ),
      }}
    />
  );
};
