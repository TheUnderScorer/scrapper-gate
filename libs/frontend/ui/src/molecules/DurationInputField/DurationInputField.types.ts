import { TextFieldProps } from '@mui/material';
import { DurationInput, Duration } from '@scrapper-gate/shared/schema';

export interface DurationInputFieldProps
  extends Omit<TextFieldProps, 'value' | 'onChange'> {
  value?: DurationInput | Duration;
  onChange?: (value: DurationInput) => unknown;
}
