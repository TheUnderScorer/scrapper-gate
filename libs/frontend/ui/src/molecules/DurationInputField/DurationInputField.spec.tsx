import { Duration, toDisplayText } from '@scrapper-gate/shared/common';
import { DurationUnit } from '@scrapper-gate/shared/schema';
import { act, render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { DurationInputField } from './DurationInputField';

describe('<DurationInputField />', () => {
  it('should render without crashing', () => {
    const cmp = render(<DurationInputField />);

    expect(cmp).toMatchSnapshot();
  });

  it('should handle value change in selected duration', () => {
    const onChange = jest.fn();
    const initialDuration = Duration.fromHours(1);

    const cmp = render(
      <DurationInputField
        value={initialDuration}
        onChange={onChange}
        label="Duration"
      />
    );

    const input = cmp.container.querySelector<HTMLInputElement>('input');
    const select = cmp.container.querySelector<HTMLInputElement>(
      '.MuiSelect-nativeInput'
    );

    expect(input?.value).toEqual('1');
    expect(select?.value).toEqual(DurationUnit.Hours);
  });

  it('should not change input value after duration change', () => {
    const onChange = jest.fn();
    const initialDuration = Duration.fromHours(1);

    const cmp = render(
      <DurationInputField
        value={initialDuration}
        onChange={onChange}
        label="Duration"
      />
    );

    act(() => {
      userEvent.click(cmp.getByText(toDisplayText(DurationUnit.Hours)));
    });

    act(() => {
      userEvent.click(cmp.getByText(toDisplayText(DurationUnit.Seconds)));
    });

    expect(onChange).toHaveBeenCalledTimes(1);
    expect(onChange).toHaveBeenCalledWith(Duration.fromSeconds(1));
  });
});
