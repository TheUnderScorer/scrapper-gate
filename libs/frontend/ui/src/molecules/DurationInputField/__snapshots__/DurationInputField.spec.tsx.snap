// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DurationInputField /> should render without crashing 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="MuiFormControl-root MuiTextField-root duration-input-text-field css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
      >
        <div
          class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-adornedEnd css-o9k5xi-MuiInputBase-root-MuiOutlinedInput-root"
        >
          <input
            aria-invalid="false"
            class="MuiOutlinedInput-input MuiInputBase-input MuiInputBase-inputAdornedEnd css-nxo287-MuiInputBase-input-MuiOutlinedInput-input"
            type="text"
            value="0"
          />
          <div
            class="MuiInputAdornment-root MuiInputAdornment-positionEnd MuiInputAdornment-outlined MuiInputAdornment-sizeMedium css-1laqsz7-MuiInputAdornment-root"
          >
            <div
              class="MuiInput-root MuiInput-underline MuiInputBase-root MuiInputBase-colorPrimary MuiSelect-root duration-input-field-select css-a88p61-MuiInputBase-root-MuiInput-root"
            >
              <div
                aria-expanded="false"
                aria-haspopup="listbox"
                class="MuiSelect-select MuiSelect-standard MuiInput-input MuiInputBase-input css-1rxz5jq-MuiSelect-select-MuiInputBase-input-MuiInput-input"
                role="button"
                tabindex="0"
              >
                Milliseconds
              </div>
              <input
                aria-hidden="true"
                class="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
                tabindex="-1"
                value="Milliseconds"
              />
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconStandard css-pqjvzy-MuiSvgIcon-root-MuiSelect-icon"
                data-testid="ArrowDropDownIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M7 10l5 5 5-5z"
                />
              </svg>
            </div>
          </div>
          <fieldset
            aria-hidden="true"
            class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="css-nnbavb"
            >
              <span
                class="notranslate"
              >
                ​
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="MuiFormControl-root MuiTextField-root duration-input-text-field css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
    >
      <div
        class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-adornedEnd css-o9k5xi-MuiInputBase-root-MuiOutlinedInput-root"
      >
        <input
          aria-invalid="false"
          class="MuiOutlinedInput-input MuiInputBase-input MuiInputBase-inputAdornedEnd css-nxo287-MuiInputBase-input-MuiOutlinedInput-input"
          type="text"
          value="0"
        />
        <div
          class="MuiInputAdornment-root MuiInputAdornment-positionEnd MuiInputAdornment-outlined MuiInputAdornment-sizeMedium css-1laqsz7-MuiInputAdornment-root"
        >
          <div
            class="MuiInput-root MuiInput-underline MuiInputBase-root MuiInputBase-colorPrimary MuiSelect-root duration-input-field-select css-a88p61-MuiInputBase-root-MuiInput-root"
          >
            <div
              aria-expanded="false"
              aria-haspopup="listbox"
              class="MuiSelect-select MuiSelect-standard MuiInput-input MuiInputBase-input css-1rxz5jq-MuiSelect-select-MuiInputBase-input-MuiInput-input"
              role="button"
              tabindex="0"
            >
              Milliseconds
            </div>
            <input
              aria-hidden="true"
              class="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
              tabindex="-1"
              value="Milliseconds"
            />
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconStandard css-pqjvzy-MuiSvgIcon-root-MuiSelect-icon"
              data-testid="ArrowDropDownIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M7 10l5 5 5-5z"
              />
            </svg>
          </div>
        </div>
        <fieldset
          aria-hidden="true"
          class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
        >
          <legend
            class="css-nnbavb"
          >
            <span
              class="notranslate"
            >
              ​
            </span>
          </legend>
        </fieldset>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
